File: endpoints/elixir/lib/zerodds.ex
# SPDX-License-Identifier: Apache-2.0
# Copyright 2026 ZeroDDS Contributors
#
# Native pure-Elixir endpoint SDK (ADR 0013): a from-scratch XCDR wire-core on
# the BEAM, byte-identical to the Rust core and the other SDKs. Sync (receive
# with a timeout) and async (a process/mailbox — the idiomatic OTP model).
#
# No native addon, no NIF: Elixir binaries + bitstring syntax carry the wire.

defmodule ZeroDDS.Wire do
  @moduledoc "XCDR writer/reader. Alignment relative to buffer start, cap 4."

  # --- Writer: a %{buf, endian} accumulator (endian :little | :big) ---

  def writer(endian), do: %{buf: <<>>, endian: endian}

  defp align(w, a) do
    cap = min(a, 4)
    pad = rem(cap - rem(byte_size(w.buf), cap), cap)
    %{w | buf: w.buf <> :binary.copy(<<0>>, pad)}
  end

  # le: the value already encoded little-endian; reversed for :big.
  defp put(w, a, le) do
    w = align(w, a)
    bytes = if w.endian == :big, do: rev(le), else: le
    %{w | buf: w.buf <> bytes}
  end

  defp rev(bin), do: bin |> :binary.bin_to_list() |> Enum.reverse() |> :binary.list_to_bin()

  def put_u8(w, v), do: %{w | buf: w.buf <> <<v::8>>}
  def put_u16(w, v), do: put(w, 2, <<v::little-16>>)
  def put_u32(w, v), do: put(w, 4, <<v::little-32>>)
  def put_u64(w, v), do: put(w, 4, <<v::little-64>>)
  def put_f32(w, v), do: put(w, 4, <<v::float-little-32>>)
  def put_bytes(w, b), do: %{w | buf: w.buf <> b}

  def put_string(w, s) do
    w = put_u32(w, byte_size(s) + 1)
    w = put_bytes(w, s)
    put_u8(w, 0)
  end

  def put_seq_u8(w, b) do
    w = put_u32(w, byte_size(b))
    put_bytes(w, b)
  end

  def bytes(w), do: w.buf

  # --- Reader: a %{bin, pos, endian} cursor ---

  def reader(bin, endian), do: %{bin: bin, pos: 0, endian: endian}

  defp take(r, a, n) do
    cap = min(a, 4)
    pad = rem(cap - rem(r.pos, cap), cap)
    pos = r.pos + pad
    chunk = binary_part(r.bin, pos, n)
    chunk = if r.endian == :big, do: rev(chunk), else: chunk
    {chunk, %{r | pos: pos + n}}
  end

  def get_u32(r) do
    {chunk, r} = take(r, 4, 4)
    <<v::little-32>> = chunk
    {v, r}
  end

  # --- primitives beyond get_u32 — the byte-exact inverse of the Writer ---

  def get_u8(r) do
    <<v::8>> = binary_part(r.bin, r.pos, 1)
    {v, %{r | pos: r.pos + 1}}
  end

  def get_u16(r) do
    {chunk, r} = take(r, 2, 2)
    <<v::little-16>> = chunk
    {v, r}
  end

  def get_u64(r) do
    {chunk, r} = take(r, 4, 8)
    <<v::little-64>> = chunk
    {v, r}
  end

  def get_f32(r) do
    {chunk, r} = take(r, 4, 4)
    <<v::float-little-32>> = chunk
    {v, r}
  end

  def get_string(r) do
    {n, r} = get_u32(r)
    s = binary_part(r.bin, r.pos, n - 1)
    {s, %{r | pos: r.pos + n}}
  end

  def get_seq_u8(r) do
    {n, r} = get_u32(r)
    b = binary_part(r.bin, r.pos, n)
    {b, %{r | pos: r.pos + n}}
  end
end

defmodule ZeroDDS.Endpoint do
  @moduledoc "XRCE WRITE_DATA framing (8-byte header + sample body)."

  @session_nokey 0x80
  @stream_best_effort 0x01

  def session_nokey, do: @session_nokey
  def stream_best_effort, do: @stream_best_effort

  def write_frame(session, stream, seq, sample) do
    n = byte_size(sample)
    <<session::8, stream::8, seq::little-16, 0x07::8, 0x03::8, n::little-16>> <> sample
  end

  def read_frame(frame)
      when byte_size(frame) >= 8 and binary_part(frame, 4, 1) == <<0x07>> do
    {:ok, binary_part(frame, 8, byte_size(frame) - 8)}
  end

  def read_frame(_), do: :error
end

defmodule ZeroDDS.Client do
  @moduledoc """
  Sync endpoint. `transport` is `%{deliver: fun/1, receive: fun/0}` where
  receive returns a frame binary or nil.
  """

  def new(transport) do
    %{
      transport: transport,
      session: ZeroDDS.Endpoint.session_nokey(),
      stream: ZeroDDS.Endpoint.stream_best_effort(),
      seq: 1
    }
  end

  def write(c, sample) do
    frame = ZeroDDS.Endpoint.write_frame(c.session, c.stream, c.seq, sample)
    c.transport.deliver.(frame)
    %{c | seq: rem(c.seq + 1, 0x10000)}
  end

  # One non-blocking receive: returns the sample body, or nil.
  def poll(c) do
    case c.transport.receive.() do
      nil ->
        nil

      frame ->
        case ZeroDDS.Endpoint.read_frame(frame) do
          {:ok, body} -> body
          :error -> nil
        end
    end
  end
end

defmodule ZeroDDS.AsyncReader do
  @moduledoc """
  Async endpoint, idiomatic BEAM: spawns a process that polls the transport and
  sends `{:zerodds_sample, body}` to `target`. The consumer receives in its own
  mailbox. `stop/1` tells the reader to exit.
  """

  def start(transport, target) do
    spawn(fn -> loop(transport, target) end)
  end

  def stop(pid), do: send(pid, :zerodds_stop)

  defp loop(transport, target) do
    receive do
      :zerodds_stop -> :ok
    after
      0 ->
        case transport.receive.() do
          nil ->
            Process.sleep(1)
            loop(transport, target)

          frame ->
            case ZeroDDS.Endpoint.read_frame(frame) do
              {:ok, body} -> send(target, {:zerodds_sample, body})
              :error -> :ok
            end

            loop(transport, target)
        end
    end
  end
end

defmodule ZeroDDS.MemTransport do
  @moduledoc "In-memory FIFO transport (an Agent) for tests + example."

  def new do
    {:ok, pid} = Agent.start_link(fn -> :queue.new() end)

    %{
      deliver: fn frame ->
        Agent.update(pid, fn q -> :queue.in(frame, q) end)
        true
      end,
      receive: fn ->
        Agent.get_and_update(pid, fn q ->
          case :queue.out(q) do
            {{:value, f}, q2} -> {f, q2}
            {:empty, q2} -> {nil, q2}
          end
        end)
      end
    }
  end
end

File: endpoints/ocaml/zerodds.ml
(* SPDX-License-Identifier: Apache-2.0 *)
(* Copyright 2026 ZeroDDS Contributors *)
(*
   Native pure-OCaml endpoint SDK (ADR 0013): a from-scratch XCDR wire-core,
   byte-identical to the Rust core and the other SDKs. Sync (poll) and async
   (a Thread filling a Mutex/Condition mailbox — stdlib, no Lwt/Async). *)

module Wire = struct
  type endian = LE | BE

  (* --- Writer: a Buffer + endian; alignment relative to buffer start, cap 4 --- *)

  type writer = { buf : Buffer.t; endian : endian }

  let writer endian = { buf = Buffer.create 64; endian }

  let align w a =
    let cap = min a 4 in
    let pad = (cap - (Buffer.length w.buf mod cap)) mod cap in
    for _ = 1 to pad do Buffer.add_char w.buf '\000' done

  (* le: the value already encoded little-endian; reversed for BE. *)
  let put w a (le : bytes) =
    align w a;
    let n = Bytes.length le in
    if w.endian = BE then
      for i = n - 1 downto 0 do Buffer.add_char w.buf (Bytes.get le i) done
    else Buffer.add_bytes w.buf le

  let le_of_int v n =
    let b = Bytes.create n in
    for i = 0 to n - 1 do
      Bytes.set b i (Char.chr ((v lsr (8 * i)) land 0xff))
    done;
    b

  let put_u8 w v = Buffer.add_char w.buf (Char.chr (v land 0xff))
  let put_u16 w v = put w 2 (le_of_int v 2)
  let put_u32 w v = put w 4 (le_of_int v 4)

  let put_u64 w (v : int64) =
    let b = Bytes.create 8 in
    for i = 0 to 7 do
      let byte = Int64.to_int (Int64.logand (Int64.shift_right_logical v (8 * i)) 0xffL) in
      Bytes.set b i (Char.chr byte)
    done;
    put w 4 b

  let put_f32 w (v : float) =
    let bits = Int32.bits_of_float v in
    let b = Bytes.create 4 in
    for i = 0 to 3 do
      let byte = Int32.to_int (Int32.logand (Int32.shift_right_logical bits (8 * i)) 0xffl) in
      Bytes.set b i (Char.chr byte)
    done;
    put w 4 b

  let put_bytes w (b : bytes) = Buffer.add_bytes w.buf b

  let put_string w s =
    put_u32 w (String.length s + 1);
    Buffer.add_string w.buf s;
    put_u8 w 0

  let put_seq_u8 w (b : bytes) =
    put_u32 w (Bytes.length b);
    Buffer.add_bytes w.buf b

  let bytes w = Buffer.to_bytes w.buf

  (* --- Reader: a bytes cursor --- *)

  type reader = { data : bytes; mutable pos : int; rendian : endian }

  let reader data endian = { data; pos = 0; rendian = endian }

  let take r a n =
    let cap = min a 4 in
    let pad = (cap - (r.pos mod cap)) mod cap in
    r.pos <- r.pos + pad;
    let b = Bytes.sub r.data r.pos n in
    r.pos <- r.pos + n;
    if r.rendian = BE then begin
      let out = Bytes.create n in
      for i = 0 to n - 1 do Bytes.set out i (Bytes.get b (n - 1 - i)) done;
      out
    end
    else b

  let get_u32 r =
    let b = take r 4 4 in
    Char.code (Bytes.get b 0)
    lor (Char.code (Bytes.get b 1) lsl 8)
    lor (Char.code (Bytes.get b 2) lsl 16)
    lor (Char.code (Bytes.get b 3) lsl 24)

  (* --- primitives beyond get_u32 — the byte-exact inverse of the Writer --- *)

  let get_u8 r =
    let v = Char.code (Bytes.get r.data r.pos) in
    r.pos <- r.pos + 1;
    v

  let get_u16 r =
    let b = take r 2 2 in
    Char.code (Bytes.get b 0) lor (Char.code (Bytes.get b 1) lsl 8)

  let get_u64 r =
    let b = take r 4 8 in
    let v = ref 0L in
    for i = 0 to 7 do
      v :=
        Int64.logor !v
          (Int64.shift_left (Int64.of_int (Char.code (Bytes.get b i))) (8 * i))
    done;
    !v

  let get_f32 r = Int32.float_of_bits (Int32.of_int (get_u32 r))

  let get_string r =
    let n = get_u32 r in
    let s = Bytes.sub_string r.data r.pos (n - 1) in
    r.pos <- r.pos + n;
    s

  let get_seq_u8 r =
    let n = get_u32 r in
    let b = Bytes.sub r.data r.pos n in
    r.pos <- r.pos + n;
    b
end

module Endpoint = struct
  let session_nokey = 0x80
  let stream_best_effort = 0x01

  let write_frame session stream seq (sample : bytes) =
    let n = Bytes.length sample in
    let hdr = Bytes.create 8 in
    Bytes.set hdr 0 (Char.chr session);
    Bytes.set hdr 1 (Char.chr stream);
    Bytes.set hdr 2 (Char.chr (seq land 0xff));
    Bytes.set hdr 3 (Char.chr ((seq lsr 8) land 0xff));
    Bytes.set hdr 4 (Char.chr 0x07);
    Bytes.set hdr 5 (Char.chr 0x03);
    Bytes.set hdr 6 (Char.chr (n land 0xff));
    Bytes.set hdr 7 (Char.chr ((n lsr 8) land 0xff));
    Bytes.cat hdr sample

  let read_frame (frame : bytes) =
    if Bytes.length frame >= 8 && Char.code (Bytes.get frame 4) = 0x07 then
      Some (Bytes.sub frame 8 (Bytes.length frame - 8))
    else None
end

(* transport: deliver a frame, receive a frame (or None). *)
type transport = { deliver : bytes -> unit; receive : unit -> bytes option }

module Client = struct
  type t = { transport : transport; session : int; stream : int; mutable seq : int }

  let create transport =
    { transport; session = Endpoint.session_nokey; stream = Endpoint.stream_best_effort; seq = 1 }

  let write c (sample : bytes) =
    let frame = Endpoint.write_frame c.session c.stream c.seq sample in
    c.transport.deliver frame;
    c.seq <- (c.seq + 1) land 0xffff

  (* One non-blocking receive: returns the sample body, or None. *)
  let poll c =
    match c.transport.receive () with
    | None -> None
    | Some frame -> Endpoint.read_frame frame
end

(* A blocking FIFO mailbox (Mutex + Condition), the idiomatic stdlib primitive. *)
module Mailbox = struct
  type 'a t = { mutable q : 'a list; m : Mutex.t; c : Condition.t }

  let create () = { q = []; m = Mutex.create (); c = Condition.create () }

  let put mb x =
    Mutex.lock mb.m;
    mb.q <- mb.q @ [ x ];
    Condition.signal mb.c;
    Mutex.unlock mb.m

  let take mb =
    Mutex.lock mb.m;
    while mb.q = [] do Condition.wait mb.c mb.m done;
    let x = List.hd mb.q in
    mb.q <- List.tl mb.q;
    Mutex.unlock mb.m;
    x
end

module AsyncReader = struct
  type t = { mutable running : bool; mb : bytes Mailbox.t }

  let start transport =
    let r = { running = true; mb = Mailbox.create () } in
    let rec loop () =
      if r.running then begin
        (match transport.receive () with
         | None -> Thread.delay 0.001
         | Some frame -> (
             match Endpoint.read_frame frame with
             | Some body -> Mailbox.put r.mb body
             | None -> ()));
        loop ()
      end
    in
    ignore (Thread.create loop ());
    r

  let recv r = Mailbox.take r.mb
  let stop r = r.running <- false
end

module MemTransport = struct
  let create () =
    let q = ref [] in
    let m = Mutex.create () in
    {
      deliver = (fun frame ->
        Mutex.lock m;
        q := !q @ [ frame ];
        Mutex.unlock m);
      receive = (fun () ->
        Mutex.lock m;
        let r = match !q with [] -> None | x :: xs -> q := xs; Some x in
        Mutex.unlock m;
        r);
    }
end
